### DISCUSSION TITLE
Release 1.2

### DISCUSSION THREAD

**Discussion by Tronic** [performous/performous#665] [General]
How about bumping the version number and pushing another release out? The current one, 1.1, is five years old.

**Comment by Baklap4** [comment]
There are still some things id like to get merged ill get back to here next wednesday

**Comment by Baklap4** [comment]
So i'd get back to you @Tronic 

### Release 1.2.0

Some things i'd still like to be within the new release:

- [x] Windows Build as this fixes lots of issues with the crosscompiled version. (#650)
- [x] ffmpeg in thread as this seemed serious (#664)
- [x] Keypress bug (#663)
- [x] Generating a deb package (#661)
- [x] Waveorder of the pitch (#653)
- [x] Audio device screen fixes (#655)
- [x] Autodetect webcam (#652)
- [x] Pitchdetector (#540) 
- [x] Runtime change language (#524)

These are all open PR's which add some nice new features and improvements / bugfixes
Most of the PR's just need some review work done.

Once those are merged i'd like to propose a beta build for Performous 1.2.0 for every platform: Windows, osx, linux.
This build will be tagged and only hotfixes will be applied to this. If all is alright after like 1 month we can release the stable 1.2 at that time.

### Release 2.0.0

As for a future release 2.0.0 there are some more things to be done:

- [ ] Rework of text positioning (#533)
- [ ] Switch of webserver dependency (#545)
- [ ] Singleton refactor (#648)
- [ ] New features
    - [ ] pitchdetector2 (#541)
    - [ ] Memorise player name (#538)
    - [ ] Completed translation (#526)
        - [x] Dutch (@Baklap4)
        - [x] French (@OznOg)
        - [ ] Spanish (@Lord-Kamina)
        - [ ] German (@TobiX)
        - [ ] Slovak (@jose1711)
    - [ ] Better theming
        - [ ] A few new themes aside from the default one
    - [ ] Other not specified features
  
How does this plan looks like?

**Comment by basisbit** [comment]
I'd suggest doing one release per year (unless there are no improvements or bugfixes). Then you can use the year number as main version number, so it is easy for people to see if they have the current version of the game installed. A yearly release also resolves the issue of the fear of big main version number changes and also gets people in the wild to "test" stuff in production, so you have a somewhat devent feedback loop, instead of only hearing feedback to breaking changes like 4 years later or so.

**Reply by Baklap4** [reply]
Within the performous tree we always used "semver"-like notation (x.y.z) up to 1.1 somehow we didn't add the patch number there. I don't think we'll change this. The idea of a yearly release is nice but doesn't add quite much value imho. At least not for a game. For an OS on the otherhand like ubuntu you can count on when to update and such.

As for newer releases i'd like to be able to release more often with smaller changes. The master branch should be stable at all times, but we all know that's not easy to manage. Maybe even a version bump per new PR but i'll have to think about that.

As for the feedback loop. Within each pr we test on several platforms already. Yes some things do get unnoticed, but that also won't be solved by releasing just once a year. With the automatic creation of artifacts we also have a way to generate nightly builds (we already do)  which is how people can test our bleeding edge. It'd be nice if we can create an automaticly deploy within debian/ubuntu repo's and/or add a way to auto-update the game. But that's out of bounds of this discussion.

**Reply by basisbit** [reply]
The major reason for year as version number and at least one release per year (as long as there are changes/improvements) is, that it kind of forces you to accomplish a planned release within that year, and it reduced maintenance overhead. This is not just a thing for operating systems, but instead it also works perfectly for free/libre open source software games.

**Comment by OznOg** [comment]
I think that we will always have "things to fix" so IMHO, we should release what we have (which is mainly ok) and target another release with the features/fixes you mentioned. Releasing quite often shows the project is alive and allows to get "quick" feedbacks on updates.

**Reply by Baklap4** [reply]
> I think that we will always have "things to fix" so IMHO

That's true, but the things listed for the 1.2 release above all have open PR's. The only thing is that they either still need to be tested or reviewed which shouldn't take that long.

> Releasing quite often shows the project is alive and allows to get "quick" feedbacks on updates.

Agreed

**Reply by basisbit** [reply]
"shoudln't take that long" was already the case 2 years ago and also one year ago. Setting deadlines when you will definitely try to publish a release definitely helps. And releasing in regular intervals helps people to be prepared for it and they know roughly in what timeframe stuff has to be done, or it will be moved to the next release.

**Reply by Baklap4** [reply]
Stuff always comes up, having deadlines or regular intervals won't make that go away.
Lately performous got some traction again with the developers. But releasing it currently as is won't be a stable release.

If we want a planning/roadmap sure that can be set up. Will it be achieved.. well only time shows.

**Reply by twollgam** [reply]
I would like to see 2 or 3 releases per year. Need not to be major releases IMHO.

**Reply by Lord-Kamina** [reply]
When you put it like that, @basisbit, I agree with the sentiment. I think already the problem we're having is the ridiculous inertia of being over 1000 commits past our last release, so we were kind of going on the "still not good enough" logic.

**Reply by Baklap4** [reply]
I agree that is/was the focus. But i think once we released 1.2.0 and we get a pr for a new functionality/bugfix/refactor we should just release a new version asap instead of waiting till a desired date and/or 'good enough' state. Ideally this should be done automaticly once merged to master as that should be stable code.

**Reply by Tronic** [reply]
I have had good experiences with fixed quaterly release dates on other projects. Everyone knows when they need to get their PRs finished if they want it on the next release, so there is a crunch when the deadline nears. This also apparently speeds up development and avoids PRs getting stuck in unmerged and possibly unfinished state.

**Reply by Baklap4** [reply]
The problem we're facing is not perse choosing a moment to release i think. We all want to release new versions. So if we release quarterly, yearly or every merge to master is irrelevant.

The problem we're actually facing is that we think the current state is not good enough to be releasable.
This is partly solved by CI which makes sure the build always works.
If we want more certaintity we should consider creating tests.
But that's a whole another subject.
Then there's also the manual testing we're doing. Currently we have a dedicated tester (shoutout to @ooshlablu for doing a great job🎉)  which just takes time, but needs to be done to ensure we don't suddenly break stuff.
There's also the PR reviews which actually is the main showstopper currently taking ages as mostly i'm the only reviewer. 
I've tried to add some structure to our PR page to what items can be tested and reviewed, now people just need to adapt it and pick it up.

The other problem we're currently facing is we're having 1000 commits on current master without a release. Which is quite alot. This resolves itself once we released 1.2.0 somewhere around 12th of october.

**Reply by basisbit** [reply]
How about publishing a release candidate / beta on 2021-09-26? A fixed date that can't be moved, a couple weeks before the planned 1.2.0 release date, a few weeks from now, and no promise of it being "stable"...

**Reply by Baklap4** [reply]
> How about publishing a release candidate / beta on 2021-09-26? A fixed date that can't be moved, a couple weeks before the planned 1.2.0 release date, a few weeks from now, and no promise of it being "stable"...

According to the post below we'll make a beta/rc at 21th september. The testing period is from that till 12th of october

**Comment by Baklap4** [comment]
### Roadmap Performous 1.2.0

So a few things have are listed as shown in https://github.com/performous/performous/discussions/665#discussioncomment-1260324
Now for a planning to when this will be achieved.

The big ones are: #650 #524 and #540
The other are minor and mostly came from the windows native build feature.

I hope to get the big ones merged within the next 2 weeks (enddate: **_14th september_**) and the little ones in the week after (enddate: **_21th september_**). If it's faster that's even better. The testing period will be between 2-3 weeks (enddate: **_12 october_**). So ideally we'll have a release by the end of next month or half into the next month.

### Roadmap Performous 2.0.0

As listed above there are coming some major changes to Performous. Some noticable some less noticable.
The feature list isn't complete as of yet thus planning a date is not quite realistic. But the focus will be on the text positioning, refactoring and webserver.

### New releases

It might be that there will be some newer releases between 1.2.0 and 2.0.0
Idealy we want to release as quick as possible when a new feature is out or a bug has been fixed. So we'll think of that in te mean time how to approach this.

**Comment by ooshlablu** [comment]
Are we going to tag master with betas and release candidates during the testing period?

**Reply by Baklap4** [reply]
To be fair what we've done in the past was just create a release, and test that. If it was good enough after a few days of testing we made a tag on that commit.

As for the upcoming 1.2.0 i'd like to do the following (option A):
1. Create a branch from master called `1.2.0-rc1`
2. Bump the version so it's actually `1.2.0-rc1`
3. Create NSIS installer for windows, deb for debian, rpm for fedora and tar.gz for other systems
4. Create a tag `1.2.0-rc1`
5. Distribute the release candidate through discord and Github Discussions and Github Releases
6. Collect feedback & bugs
7. If there need something to be changed commit on the branch `1.2.0-rc1` AND master
8. If we've collected enough 'fixes' try a `1.2.0-rc2` in the same steps.
9. If no feedback & bugs are collected anymore merge the latest `rc` branch to master and tag the release on a final locked branch `1.2.0`

Another way (option B) would be to freeze master and only do rc commits on it. This means no other PR's can get merged till after the release

Which option do you guys want? React with emoji's on this below
Option A: 👍
Option B: 🎉

**Reply by Lord-Kamina** [reply]
Going with A, so long as we're understood commits to master not pertaining to bugs will not suddenly be merged into the RC.

**Reply by Baklap4** [reply]
Well in both scenario's sudden bugs won't appear as in scenario B master gets frozen until we have a release.. But yeah this means current development on those pr's will not get merged within that testing timespan which isn't quite nice :p I think we'll go with option A, but let's see what the others say about it

**Comment by Baklap4** [comment]
Alright as you guys can see the checklist for 1.2.0 is done and before schedule.
I was wondering if we could add another point to it, which is translating (or making them complete) the game in the current languages we have. So we've got a nice localized game instead of having bits english bits native tongue. **The deadline is still 21st of September!**

A start has already been made in #526

Please react with 👍 if you guys want this within the 1.2.0 release or a 👎 if you guys dont want it in the upcoming release.
Do note: We can change language on runtime now, so it's easier to switch between them without having to kill the game.

**Reply by twollgam** [reply]
I would vote up if translation work stops on 21th september so that the release 1.2 will not be later then 12 october.

BTW I can help doing german translation.

**Reply by Baklap4** [reply]
That's for sure the goal yes! Beta's will start on that date. 
Translations is just a textfile which imho can get added as 'hotfix' on the branch itself, but i doubt if that's necessairy

**Comment by Baklap4** [comment]
Locking this thread since we've just released 1.2.0-rc1 -> https://github.com/performous/performous/discussions/685
